### ISSUE TITLE
Support for special characters

### ISSUE THREAD

**Issue by fireflames** [pyepics/pyepics#87]
Hey,

for me it seems like pyepics has some trouble parsing strings with special characters.
Example:
caget('String_PV') returns "J\303\274rgen" which translates to "Jürgen", since \303 \274 is utf8 encoded "ü" in octal (hexadecimal would be "c3 bc")

Doing the same with pyepics returns neither the octal version nor the "ü", but instead: "JÃ¼rgen". Is there a way to get the same result as for the epics base caget()? Or even the correctly translated special character?

**Comment by newville** [comment]
@fireflames The behavior might be different for Python 2 and 3.  Which are you using?

There is not much intentional parsing or manipulation of strings going on.   For Python2, pyepics basically just returns `str(result)` on what it receives from the ca_get() from the C library.  For Python 3, that becomes `str(result, 'latin1')`.    Perhaps 'latin_1' is a poor default, or perhaps that encoding should be 'utf8' and/or user configurable? 

With Python 3.6.1 (Anaconda Python, on Darwin), I definitely see differences between 'latin1' and 'utf8', unrelated to Epics.  With this as `stest.py`:

    name1 = "J\303\274rgen"
    print("Version 1: ", name1, [ord(i) for i in name1])
    name2 = "Jürgen"
    print("Version 2: ", name2, [ord(i) for i in name2])

I see this:

    ~>export PYTHONIOENCODING=utf8
    ~>python stest.py
    Version 1:  JÃ¼rgen [74, 195, 188, 114, 103, 101, 110]
    Version 2:  Jürgen [74, 252, 114, 103, 101, 110]
    ~>export PYTHONIOENCODING=latin-1
    ~>python stest.py
    Version 1:  Jürgen [74, 195, 188, 114, 103, 101, 110] 
    Version 2:  J?rgen [74, 252, 114, 103, 101, 110]

From this, I think the issue may have as much to do with which encoding is used as what the pyepics library is doing.   This test also seems to imply that your version of the string is encoded using 'latin1'.  But, I'm certainly not an expert on Unicode, and perfectly willing to be corrected on any of these points.

Perhaps the right way forward is for pyepics to use PYTHONIOENCODING, as with

    EPICS_STR_ENCODING = os.environ.get('PYTHONIOENCODING', 'utf8')

in `utils3.py`.   Any thoughts, suggestions?

**Comment by ralphlange** [comment]
Just for the record:
EPICS and CA are completely encoding agnostic. Strings are just arrays of bytes.
Which is bad by itself (string operations will happily tear apart multi-byte characters), but means that clients should ultimately be configurable to be able to handle any context situation.
Nowadays, UTF-8 seems like a reasonable default.

**Comment by newville** [comment]
I changed `utils3.py` to have
```python:
EPICS_STR_ENCODING = os.environ.get('PYTHONIOENCODING', 'utf-8')
```
and I think we're very close to a release for 3.2.7

@ralphlange: it is hard to disagree that a) 'utf-8' is a reasonable default encoding, and b) that Epics follows the very old C / Unix convention that a string can be represented by 'char *'  that represents some extension of ASCII.   But: EPICS and CA are not "encoding agnostic", they are "encoding blind". 

A string is not just an array of bytes. It is an array of bytes PLUS AN ENCODING.  

One might argue that CA's use of 'char *' as string was slightly forgivable in the mid 1990s and that space efficiency was more important at that time.   Well, except that an EPICS string exists primarily to be shared between computers, and an EPICS string record does have a number of extra fields: count, type, host, access, status, and so on.  What it does not have is "encoding". 

The sender and receiver of all Epics strings MUST agree on the encoding, and there is no mechanism available for communicating that encoding.   This is a common mistake to make, but it is still a mistake.

Expecting "the client" to be able to handle any situation is not possible.  A client will likely expect a consistent encoding for all strings.  For example, pyepics now expects one encoding per process: there is no easy way to have it more fine-grained than that.  If different servers use different encodings, any client will be sometimes be wrong.  It's hard to blame the server -- it uses a consistent encoding.  It is hard to blame the client -- it also uses a consistent encoding and has no information from any server about what encoding is used.  Instead, it is a fault of the protocol and of the CA library to assert that "strings are just arrays of bytes" when they are not.

**Comment by ralphlange** [comment]
I couldn't agree more.

The next generation network protocol pvAccess is slightly better by [requiring UTF-8](http://epics-pvdata.sourceforge.net/pvAccess_Protocol_Specification.html#dataEncoding:strings) for all strings.

As for your EPICS database ideas: EPICS Base development is not free. As soon as such a request gets funded by contributing appropriate manpower or budget, it will be implemented. Meanwhile it would be helpful to note it in an [EPICS Base issue](https://bugs.launchpad.net/epics-base).

**Comment by newville** [comment]
@fireflames  Should be better (if not fixed) in 3.2.7

**Comment by fireflames** [comment]
Yes, this fixed the problem for me, thanks!
